### PR TITLE
[preview] prefilter engines v76

### PR DESCRIPTION
**Pull request by inliniac**
Changes since #2197:
- rebase to master
- make 'packet header engines' optional: --set detect.prefilter.default=auto to enable, or explicitly use 'prefilter' keyword in rules
- add additional logic for prefilter engines for rules that match on special src/dst ports or alproto
- revert the removal of the syn specific non-mpm/non-prefilter list
- various performance optimizations and profiling options

Note: the profiling code itself is quite expensive

Prscript:
- PR inliniac-pcap: https://buildbot.openinfosecfoundation.org/builders/inliniac-pcap/builds/476
- PR inliniac: https://buildbot.openinfosecfoundation.org/builders/inliniac/builds/481

### CODE REVIEW

**Comment by pevma** [comment]
Maybe a comment with the available options in yaml - "mpm/auto" and what they do or how they differ here - 
https://github.com/inliniac/suricata/pull/2207/files#diff-90

**Comment by inliniac** [comment]
Yeah adding that. In short the setting 'mpm' is what has been the default all along. So Suricata prefilters on MPM. The setting 'auto' looks at the rules to see if it can add prefilter engines for keywords like dsize, flags, flow, icode, itype, etc.

File: src/detect-engine-prefilter.c [hunk 1]
```diff
@@ -58,6 +58,31 @@
 static int PrefilterStoreGetId(const char *name);
 #endif
 
+/* Return true is the list is sorted smallest to largest */
```
Comment (gozzy) [inline src/detect-engine-prefilter.c]: Hmm... It doesn't seem to return anything ;)

**Comment by inliniac** [comment]
Replaced by #2212

File: src/detect-engine-prefilter.c (hunk 1 above)
Comment (inliniac) [inline src/detect-engine-prefilter.c]: Good point, removing.

**Pull request closed without merging**